### PR TITLE
Squiz/ClosingDeclarationComment: fix a non-problematic bug plus some other small improvements

### PR DESCRIPTION
**Pull request by rodrigoprimo**
## Description

This PR makes the following changes to the Squiz.Commenting.ClosingDeclarationComment sniff:

- Improve test coverage
- Remove an unreachable condition
- Prevent a non-problematic bug on line 96 where the code `if (rtrim($tokens[$next]['content']) === $comment)` was not taking into account that `$next` can be false. To fix it, a `$next !== false` check was added to the if condition.

@jrfnl suggested those changes while reviewing another PR: https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/340#issuecomment-1946387893. In the same comment, it was also suggested that the sniff should handle traits. This will be addressed in a separate PR.

I noticed that this sniff triggers a warning when there is a missing opening or closing bracket. I imagine this could be misleading when using PHPCS with live code, and it is the first sniff that I see doing this. I'm highlighting this behavior in case we want to address it in another PR. For now, I added two tests documenting this behavior.

## Suggested changelog entry

Squiz.Commenting.ClosingDeclarationComment: fixed a non-problematic bug when $next is false

## Related issues/external references

Related to https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/340#issuecomment-1946387893 and #146

## Types of changes

- [x] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement

## PR checklist

- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] [Required for new sniffs] I have added XML documentation for the sniff.

### CODE REVIEW

**Comment by jrfnl** [comment]
> I noticed that this sniff triggers a warning when there is a missing opening or closing bracket. I imagine this could be misleading when using PHPCS with live code, and it is the first sniff that I see doing this. I'm highlighting this behavior in case we want to address it in another PR. For now, I added two tests documenting this behavior.

@rodrigoprimo There are currently a small number of sniffs which do this and this behaviour will be removed in PHPCS 4.0 as per https://github.com/squizlabs/PHP_CodeSniffer/issues/2455

File: src/Standards/Squiz/Tests/Commenting/ClosingDeclarationCommentUnitTest.4.inc [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?php
```
Comment (jrfnl) [inline src/Standards/Squiz/Tests/Commenting/ClosingDeclarationCommentUnitTest.4.inc:1]: While this is a good test to have anyway, this test doesn't actually highlight the problem with the missing `$next !== false` condition.

We looked at this during our call and the following additional test should be added to safeguard the fix:
```php
//end closingBraceAtEndOfFileMissingComment()

<?php

function closingBraceAtEndOfFileMissingComment() {
}
```

This test will show that without that extra condition the _wrong_ error is being thrown (though an error is still being thrown).


_Note: as the test framework as-is doesn't safeguard that the correct error code is being thrown for each line, it's not a hard safeguard, but having the test in place will allow us to safeguard this properly if/when the test framework gets updates to check per error code/per line._

**Review by jrfnl — CHANGES_REQUESTED** [review]
Hi @rodrigoprimo Sorry for taking so long before reviewing this PR properly.

We've now looked at it together in a call, so you are already aware of the remarks I'm making.

Overall: looking good! Just some small changes needed to finish this off properly.

**Comment by rodrigoprimo** [comment]
Thanks for the review, @jrfnl!

I just pushed a few new commits implementing all the changes that you suggested and this PR is ready for a final review.

**Comment by jrfnl** [comment]
Rebased without changes, just some reorganization of the commits. Merging once the build passes.

**Pull request merged**